### PR TITLE
implement eq_dyn, neq_dyn, lt_dyn, lt_eq_dyn, gt_dyn, gt_eq_dyn for timestamp types

### LINKED ISSUE #1086: implement `eq_dyn`, `neq_dyn`, `lt_dyn`, `lt_eq_dyn`, `gt_dyn`, `gt_eq_dyn` for timestamp types
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
We are tryng to use the new dynamic comparison kernels added in https://github.com/apache/arrow-rs/pull/858 / #843 by @Jimexist 
 in DataFusion: https://github.com/apache/arrow-datafusion/pull/1475

It turns out that Datafusion also supports operations on the timestamp / time arrays as well so we can't use the dynamic kernels (yet)

**Describe the solution you'd like**
Add support for the following types to `eq_dyn`, `neq_dyn`, `lt_dyn`, `lt_eq_dyn`, `gt_dyn`, `gt_eq_dyn` (there is one macro needed):
```
DataType::Timestamp(TimeUnit::Nanosecond, _) => {
DataType::Timestamp(TimeUnit::Microsecond, _) => {
DataType::Timestamp(TimeUnit::Millisecond, _) => {
DataType::Timestamp(TimeUnit::Second, _) => {
DataType::Date32 => {
DataType::Date64 => {
```

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1086.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by Jimexist — APPROVED** [review]

**Review by liukun4515 — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Thanks for review. Is this ready to go?

File: arrow/src/compute/kernels/comparison.rs [hunk 1]
```diff
@@ -1496,6 +1553,16 @@ mod tests {
         };
     }
 
+    /// Evaluate `KERNEL` with two vectors as inputs and assert against the expected output.
+    /// `A_VEC` and `B_VEC` can be of type `Vec<i64>` or `Vec<Option<i64>>`.
+    /// `EXPECTED` can be either `Vec<bool>` or `Vec<Option<bool>>`.
+    /// The main reason for this macro is that inputs and outputs align nicely after `cargo fmt`.
+    macro_rules! cmp_i64 {
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:1560]: this is a nice refactor

**Review by alamb — APPROVED** [review]
Thank you @viirya

**Comment by viirya** [comment]
Thanks @alamb @Jimexist @liukun4515 !

**Pull request merged**